src/js/historical.cpp

```diff
@@ -23,8 +23,34 @@ namespace ccf::js
         js_receipt,
         "cert",
         JS_NewString(ctx, receipt_out.cert.value().c_str()));
-    JS_SetPropertyStr(
-      ctx, js_receipt, "leaf", JS_NewString(ctx, receipt_out.leaf.c_str()));
+    if (receipt_out.leaf.has_value())
+    {
+      JS_SetPropertyStr(
+        ctx, js_receipt, "leaf", JS_NewString(ctx, receipt_out.leaf->c_str()));
+    }
+    if (receipt_out.leaf_components.has_value())
+    {
+      auto leaf_components = JS_NewObject(ctx);
+      if (receipt_out.leaf_components->write_set_digest.has_value())
+      {
+        JS_SetPropertyStr(
+          ctx,
+          leaf_components,
+          "write_set_digest",
+          JS_NewString(
+            ctx, receipt_out.leaf_components->write_set_digest->c_str()));
+      }
+      if (receipt_out.leaf_components->claims_digest.has_value())
```

Comment:
When can this be missing?

Reply:
Same answer as above. We could have an InternalTxReceipt, strictly for interface between the C++ historical and JS historical ifaces I guess, where the leaf_components aren't optional, rather than reuse the same schema as for the user API, and then wrap it again for the C++ API.
It'd be quite a lot of extra wrappers to save two conditions though.